activesupport/lib/active_support/key_rotator.rb

```diff
@@ -0,0 +1,210 @@
+  #
+  #   legacy_scheme = {
+  #     cipher: :aes_cbc_hmac, salt: 'encrypted cookie', signed_salt: 'signed encrypted cookie'
+  #   }
+  #
+  #   rotator = ActiveSupport::KeyRotator.new(key, scheme, legacy_scheme)         # => <ActiveSupport::KeyRotator ...>
+  #   rotator.decrypt_and_verify(rotator.encrypt_and_sign('some secret message')) # => 'some secret message'
+  #
+  #   # Using a message encrypted with the legacy scheme
+  #   rotator.decrypt_and_verify(legacy_message)                                  # => 'some legacy encrypted message'
+  #
+  #   # MessageEncryptor::InvalidMessage will be raised if none of the wrapped
+  #   # encryptors are able to decrypt the message
+  #   rotator.decrypt_and_verify(unknown_message)                                 # => <ActiveSupport::MessageEncryptor::InvalidMessage>
+  class KeyRotator
+    class UnsupportedScheme < StandardError; end
+
+    AES_GCM_CIPHER  = "aes-256-gcm".freeze
+    AES_GCM_KEY_LEN = MessageEncryptor.key_len(AES_GCM_CIPHER).freeze
+
+    AES_CBC_HMAC_CIPHER  = "aes-256-cbc".freeze
+    AES_CBC_HMAC_KEY_LEN = MessageEncryptor.key_len(AES_CBC_HMAC_CIPHER).freeze
+
+    module KeyGenMethods
+      private
+
+        def validate_key_generator!
+          if LegacyKeyGenerator === @primary_key_generator
+            raise "You didn't set secrets.secret_key_base, which is required for this class. " \
+              "Read the upgrade documentation to learn more about this new config option."
+          end
+        end
+
+        def key_generator(secret)
+          return @primary_key_generator unless secret
+
+          KeyGenerator.new(secret, iterations: 1000)
+        end
+    end
+
+    include KeyGenMethods
+
+    # Initialize a new KeyRotator. A primary <tt>ActiveSupport::KeyGenerator</tt>
+    # and scheme must be supplied. One or more legacy schemes may also be
+    # supplied.
+    #
+    # A scheme should define the <tt>:cipher</tt> and <tt>:salt</tt> value. An
+    # optional <tt>:secret</tt> is also accepted. When supplied, a new
+    # KeyGenerator will created. This value is intended to allow rotating of
+    # the <tt>:secret_key_base</tt> application value.
+    #
+    # The valid <tt>:cipher</tt> values are: <tt>:aes_gcm</tt>, <tt>:aes_cbc_hmac</tt>,
+    # and <tt>:secret_token</tt>. Only <tt>:aes_gcm</tt> may be used for the
+    # primary scheme.
+    #
+    # Additionally, when using the <tt>:aes_cbc_hmac</tt>, a second
+    # <tt>:signed_salt</tt> value must be supplied.
+    def initialize(primary_key_generator, primary_scheme, *legacy_schemes)
+      @primary_key_generator = primary_key_generator
+
+      validate_key_generator!
+
+      @primary_encryptor = encryptor_from_scheme(primary_scheme)
+      @legacy_encryptors = legacy_schemes.map { |os| encryptor_from_scheme(os) }
+    end
+
+    delegate :encrypt_and_sign, :decrypt_and_verify, to: :@primary_encryptor
+
+    def decrypt_and_verify_with_legacy_encryptors(value)
```

Comment:
This is how I originally designed the `KeyRotator` class. The problem was, in the cookie middleware `parse` methods, we need to know when the primary encryptor/verifier fails so that the cookie can be updated.

I experimented with having `decrypt_and_verify` that accepts an optional block that is called when the primary encryptor/verifier fails. It didn't do the trick though because you end up having to `deserialize` before doing the actual cooke update.

Definitely open to ideas on how we can have `decrypt_and_verify` and `verify` signal to their caller that primary encryptor or verifier were not used. This would definitely result in a simpler and easier to use API as you have suggested.